frozen-abi/macro/src/lib.rs

```diff
@@ -28,41 +28,55 @@ pub fn derive_abi_enum_visitor(_item: TokenStream) -> TokenStream {
 }
 
 #[cfg(RUSTC_WITH_SPECIALIZATION)]
-use proc_macro2::{Span, TokenStream as TokenStream2, TokenTree::Group};
+use proc_macro2::{Span, TokenStream as TokenStream2, TokenTree};
 #[cfg(RUSTC_WITH_SPECIALIZATION)]
-use quote::quote;
+use quote::{quote, ToTokens};
 #[cfg(RUSTC_WITH_SPECIALIZATION)]
 use syn::{
-    parse_macro_input, Attribute, AttributeArgs, Error, Fields, Ident, Item, ItemEnum, ItemStruct,
-    ItemType, Lit, Meta, NestedMeta, Variant,
+    parse_macro_input, Attribute, Error, Fields, Ident, Item, ItemEnum, ItemStruct, ItemType,
+    LitStr, Variant,
 };
 
 #[cfg(RUSTC_WITH_SPECIALIZATION)]
 fn filter_serde_attrs(attrs: &[Attribute]) -> bool {
-    let mut skip = false;
-
-    for attr in attrs {
-        let ss = &attr.path.segments.first().unwrap().ident.to_string();
-        if ss.starts_with("serde") {
-            for token in attr.tokens.clone() {
-                if let Group(token) = token {
-                    for ident in token.stream() {
-                        if ident.to_string() == "skip" {
-                            skip = true;
-                        }
+    fn contains_skip(tokens: TokenStream2) -> bool {
+        for token in tokens.into_iter() {
+            match token {
+                TokenTree::Group(group) => {
+                    if contains_skip(group.stream()) {
+                        return true;
                     }
                 }
+                TokenTree::Ident(ident) => {
+                    if ident == "skip" {
+                        return true;
+                    }
+                }
```

Comment:
nit: I'd like to avoid nested `if`s while still using this new exhaustive match.

```rust
match token {
    TokenTree::Group(group) if contains_skip(group.stream()) | TokenTree::Ident(ident) if ident == "skip" {
        return true;
    }
    TokenTree::Group(_) | TokenTree::Ident(_) | TokenTree::Punct(_) | TokenTree::Literal(_) => (),
}
```

Reply:
What is bad about the nested `if`s?

The first level does an exhaustive match on the token type.
And then for each type we make a decision.
While we can combine cases because there are only two possible outcomes, I think, it makes the code harder to read.

I think, `ident == "skip"` and `contains_skip(group.stream())` are checks that are important for the logic flow.
"Hiding" them in the guards seems counterintuitive.

Also, Rust does not group guards with the pattern branches.
Branches are combined into a single pattern, and a single guard can be provided.
So it is not syntactically correct.  All branches of a or-pattern needs to introduce the same set of bindings for the pattern to be [syntactically valid](https://doc.rust-lang.org/nightly/reference/patterns.html#static-semantics).

I can combine the last two cases together in an or-pattern.
```rust
                TokenTree::Punct(_) | TokenTree::Literal(_) => (),
```

Maybe because not all languages support or-patterns, I might not be as used to them.
But I mostly find them reducing the code readability, except for cases when the focus needs to be on the action, rather than on the value being deconstructed.
In which case they are really great.
But, in my mind, in all the other cases they just make the reader work harder to understand the code.

Reply:
got it. thanks for explaining in detail.